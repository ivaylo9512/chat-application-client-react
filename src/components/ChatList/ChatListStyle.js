import styled from 'styled-components'

export const ChatInfo = styled.div`
    color: white;
    width: 100%;
    text-align: center;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 2.5vh;
`

export const Span = styled.span`
    border-radius: 1vw;
    padding: 0 3%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    font-weight: bold;
    font-size: 0.85vw;
    position: relative;
    background: rgb(24,27, 33);
    box-shadow: 0px 0px 0.6vh 0.2vh rgb(0 0 0 / 75%);
`

export const MoreInfo = styled.span`
    width: 2.8vh;
    height: 2.8vh;
    text-align: center;
    box-shadow: 0px 0px 0.15vw 0.15vw rgb(0 0 0);
    background: rgb(24,27,33);
    border-radius: 50%;
    display: flex;
    color: white;
    align-items: center;
    justify-content: center;
    font-size: 1.5vh;
    flex: 0 0 2.8vh;
    margin-right: 1vw;
`

export const LoadingContainer = styled.div`
    width: 99%;
    display: flex;
    justify-content: center;
    top: 11vh;
    
    > div{
        width: 5vh;
        height: 5vh;
        background: rgb(24,27,33);
        box-shadow: 0px 0px 0.1vw 0.1vw rgb(0 0 0);
        border-radius: 1.5vh;
    }
`