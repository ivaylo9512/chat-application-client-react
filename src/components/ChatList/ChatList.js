import React, { useEffect } from 'react';
import ChatUser from '../Chat/Chat';
import { useSelector, useDispatch } from 'react-redux';
import { getChatsQuery, chatsRequest, getChatsState } from '../../app/slices/chatsSlice';
import { ChatInfo, Span } from './ChatListStyle'
const ChatList = () => {
    const { isLoading, data: { chats, isLastPage } } = useSelector(getChatsState);
    const dispatch = useDispatch();
    const query = useSelector(getChatsQuery);

    useEffect(() => {
        dispatch(chatsRequest({...query}));
    },[])

    return(
        <>
            {chats.length == 0 
                ? isLoading 
                    ? 'loading'
                    : <ChatInfo>
                        <Span>You don't have any chats.</Span>
                    </ChatInfo>
                : <>
                    {chats.map(chat => <ChatUser key={chat.id} chat={chat}/>)}
                    {!isLastPage && <span>{'>'}</span>}
                </>
            }   
        </>
    )

}
export default ChatList