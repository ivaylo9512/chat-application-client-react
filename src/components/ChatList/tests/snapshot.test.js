import * as redux from 'react-redux';
import { shallow } from 'enzyme';
import ChatList from 'components/ChatList/ChatList';

describe('ChatList snapshot tests', () => {
    beforeEach(() => {
        jest.spyOn(redux, 'useDispatch').mockReturnValue(jest.fn());
    })

    const createWrapper = (state) => {
        jest.spyOn(redux, 'useSelector').mockReturnValue(state);

        return shallow(<ChatList />)
    }

    it('should render Chats', () => {
        const wrapper = createWrapper({ isLoading: false, data: { chats: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 } ], isLastPage: true }});

        expect(wrapper).toMatchSnapshot();
    })

    it('should render LoadingIndicator when loading', () => {
        const wrapper = createWrapper({ isLoading: true, data: { chats: [], isLastPage: false }});

        expect(wrapper).toMatchSnapshot();
    })

    it('should render info when empty chats array and not loading ', () => {
        const wrapper = createWrapper({ isLoading: false, data: { chats: [], isLastPage: false }});

        expect(wrapper).toMatchSnapshot();
    })

    it('should render Chats and more button when lastPage is false', () => {
        const wrapper = createWrapper({ isLoading: false, data: { chats: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 } ], isLastPage: false }});

        expect(wrapper).toMatchSnapshot();
    })
})