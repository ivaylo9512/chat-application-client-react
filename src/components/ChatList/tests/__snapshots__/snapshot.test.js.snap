// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatList snapshot tests should render Chats 1`] = `
<Fragment>
  <Chat
    chat={
      Object {
        "id": 1,
      }
    }
    key="1"
  />
  <Chat
    chat={
      Object {
        "id": 2,
      }
    }
    key="2"
  />
  <Chat
    chat={
      Object {
        "id": 3,
      }
    }
    key="3"
  />
  <Chat
    chat={
      Object {
        "id": 4,
      }
    }
    key="4"
  />
</Fragment>
`;

exports[`ChatList snapshot tests should render Chats and more button when lastPage is false 1`] = `
<Fragment>
  <Chat
    chat={
      Object {
        "id": 1,
      }
    }
    key="1"
  />
  <Chat
    chat={
      Object {
        "id": 2,
      }
    }
    key="2"
  />
  <Chat
    chat={
      Object {
        "id": 3,
      }
    }
    key="3"
  />
  <Chat
    chat={
      Object {
        "id": 4,
      }
    }
    key="4"
  />
  <span
    data-testid="more"
  >
    &gt;
  </span>
</Fragment>
`;

exports[`ChatList snapshot tests should render LoadingIndicator when loading 1`] = `
<Fragment>
  <div>
    <LoadingIndicator />
  </div>
</Fragment>
`;

exports[`ChatList snapshot tests should render info when empty chats array and not loading  1`] = `
<Fragment>
  <styled.div
    data-testid="info"
  >
    <styled.span>
      You don't have any chats.
    </styled.span>
  </styled.div>
</Fragment>
`;
